File: java/src/main/java/be/steinsomers/bron_kerbosch/study/GraphTestData.java
package be.steinsomers.bron_kerbosch.study;

import be.steinsomers.bron_kerbosch.UndirectedGraph;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

record GraphTestData(UndirectedGraph graph, int cliqueCount) {
    private static List<Set<Integer>> new_sets(int n) {
        return Stream
                .generate(() -> (Set<Integer>) new HashSet<Integer>(16))
                .limit(n)
                .toList();
    }

    static GraphTestData readUndirected(String orderStr, int order, int size) throws IOException {
        assert order > 2;
        assert size >= 0;
        var fullyMeshedSize = ((long) order) * (order - 1) / 2;
        if (size > fullyMeshedSize) {
            throw new IllegalArgumentException(String.format(
                    "%d nodes accommodate at most %d edges", order, fullyMeshedSize));
        }

        var edgesPath = Paths.get("..", "data", "random_edges_order_" + orderStr + ".txt");
        var statsPath = Paths.get("..", "data", "random_stats.txt");
        var adjacencies = readEdges(edgesPath, order, size);
        var cliqueCount = readStats(statsPath, orderStr, size);

        var g = new UndirectedGraph(adjacencies);
        if (g.order() != order) throw new AssertionError("order mishap");
        if (g.size() != size) throw new AssertionError("size mishap");
        return new GraphTestData(g, cliqueCount);
    }

    private static List<Set<Integer>> readEdges(Path path, int order, int size) throws IOException {
        var adjacencies = new_sets(order);
        try (var br = Files.newBufferedReader(path)) {
            String line;
            int lineNum = 0;
            while (lineNum < size && (line = br.readLine()) != null) {
                ++lineNum;
                var fields = line.split(" ", 2);
                int v;
                int w;
                //noinspection ProhibitedExceptionCaught
                try {
                    v = Integer.parseInt(fields[0]);
                    w = Integer.parseInt(fields[1]);
                } catch (NumberFormatException | ArrayIndexOutOfBoundsException err) {
                    //noinspection ThrowInsideCatchBlockWhichIgnoresCaughtException
                    throw new IOException("File " + path + " contains bogus text " + line);
                }
                adjacencies.get(v).add(w);
                adjacencies.get(w).add(v);
            }
            if (lineNum < size) {
                throw new IOException("Exhausted list of " + lineNum + " edges in " + path);
            }
        }
        return adjacencies;
    }

    private static int readStats(Path path, String orderStr, int size) throws IOException {
        var prefix = String.format("%s\t%d\t", orderStr, size);
        try (var br = Files.newBufferedReader(path)) {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.startsWith(prefix)) {
                    try {
                        return Integer.parseInt(line.substring(prefix.length()));
                    } catch (NumberFormatException err) {
                        //noinspection ThrowInsideCatchBlockWhichIgnoresCaughtException
                        throw new IOException("File " + path + " has bogus line “" + line + "”");
                    }
                }
            }
            throw new IOException("File " + path + " lacks order " + orderStr + " size " + size);
        }
    }
}

File: kotlin/src/main/be/steinsomers/bron_kerbosch/study/GraphTestData.kt
package be.steinsomers.bron_kerbosch.study

import be.steinsomers.bron_kerbosch.UndirectedGraph
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.stream.Stream

internal data class GraphTestData(val graph: UndirectedGraph, val cliqueCount: Int) {
    companion object {
        private fun newSets(n: Int): MutableList<MutableSet<Int>> {
            return Stream
                .generate { HashSet<Int>(16) as MutableSet<Int> }
                .limit(n.toLong())
                .toList()
        }

        @Throws(IOException::class)
        fun readUndirected(orderStr: String, order: Int, size: Int): GraphTestData {
            require(order > 2)
            require(size >= 0)
            val fullyMeshedSize = (order.toLong()) * (order - 1) / 2
            require(size <= fullyMeshedSize) { "$order nodes accommodate at most $fullyMeshedSize edges" }

            val edgesPath = Paths.get("..", "data", "random_edges_order_$orderStr.txt")
            val statsPath = Paths.get("..", "data", "random_stats.txt")
            val adjacencies = readEdges(edgesPath, order, size)
            val cliqueCount = readStats(statsPath, orderStr, size)

            val g = UndirectedGraph(adjacencies)
            require(g.order == order) { "order mishap" }
            require(g.size() == size) { "size mishap" }
            return GraphTestData(g, cliqueCount)
        }

        @Throws(IOException::class)
        private fun readEdges(path: Path, order: Int, size: Int): MutableList<MutableSet<Int>> {
            val adjacencies: MutableList<MutableSet<Int>> = newSets(order)
            Files.newBufferedReader(path).use { br ->
                for (lineNum in 0..<size) {
                    val line = br.readLine()
                    if (line == null)
                        throw IOException("File $path has only $lineNum lines of the $size requested")
                    val fields: Array<String> = line.split(" ".toRegex(), limit = 2).toTypedArray()
                    val v: Int
                    val w: Int
                    try {
                        v = fields[0].toInt()
                        w = fields[1].toInt()
                    } catch (_: NumberFormatException) {
                        throw IOException("File $path contains bogus text $line")
                    } catch (_: ArrayIndexOutOfBoundsException) {
                        throw IOException("File $path contains bogus text $line")
                    }
                    adjacencies[v].add(w)
                    adjacencies[w].add(v)
                }
            }
            return adjacencies
        }

        @Throws(IOException::class)
        private fun readStats(path: Path, orderStr: String, size: Int): Int {
            val prefix = "$orderStr\t$size\t"
            var cliqueCount = 0
            Files.newBufferedReader(path).use { br ->
                br.forEachLine { line ->
                    if (line.startsWith(prefix)) {
                        if (cliqueCount != 0)
                            throw IOException("File $path multiply defines order $orderStr size $size")
                        try {
                            cliqueCount = line.substring(prefix.length).toInt()
                        } catch (_: NumberFormatException) {
                            throw IOException("File $path has bogus line “$line”")
                        }
                        if (cliqueCount <= 0)
                            throw IOException("File $path has stupid line “$line”")
                    }
                }
            }
            if (cliqueCount == 0)
                throw IOException("File $path lacks order $orderStr size $size")
            return cliqueCount
        }
    }
}
